### ISSUE TITLE
Disable autoindent

### ISSUE THREAD

**Issue by cfjedimaster** [microsoft/vscode#5446]
- VSCode Version: 1.0
- OS Version: OSX

Currently there are multiple ways to configure indention, but no way to _disable_ auto indention. For example, I enter `<p>`, hit the enter key, and on the next line the cursor has indented. I want to disable that.

**Comment by alexandrudima** [comment]
@cfjedimaster We have an open PR #5423 that would make it that pressing subsequent Enter, the inserted whitespace gets trimmed.

Also, are you finding this behaviour annoying in a certain programming language, or when doing something specific, most of the feedback we got so far is that indenting to the right spot is a helpful feature, and not an annoyance, so I'm wondering if you're hitting a use-case where this doesn't work as expected.

**Comment by cfjedimaster** [comment]
I'm finding it only an issue in HTML, maybe because I tend to have specific ways of indenting in HTML, specifically I keep head and body at the same level as html.  #5423 doesn't look to be exactly what I want - I really just want no white space inserted at all. I hit enter, and my cursor will be at position 1 in the next line.

**Comment by cfjedimaster** [comment]
Another example - given form fields where I'm wrapping each label and input in a particular div, I like to have them all at the same indentation level.

**Comment by Luminoth** [comment]
I'd like to second this feature request. Right now not being able to turn that off is the only thing stopping me from using VSCode as my main editor.

**Comment by maximecb** [comment]
Would also like to be able to disable auto indent. I generally don't like automatic behaviors that try to guess what I want to do, because they often guess incorrectly.

**Comment by peredin** [comment]
Just like others have already mentioned this "feature" is quite annoying. I dislike any kind of automatic behavior that cannot be turned off. If I want to indent a line I will use the tab key manually.

**Comment by ste2425** [comment]
Auto indenting is also quite annoying when working with projects that enforce the `no-trailing-spaces` rule of ESlint and others.

The auto indenting instantly fails this rule.

**Comment by roman01la** [comment]
I'm using Parinfer plugin for auto balancing Clojure code. Auto indentation on Enter press doesn't allow Parinfer to place a cursor into inferred position.

**Comment by Kobnar** [comment]
There's the pretty obvious use case of editing raw text.

I understand vscode is a _code_ editor, but it's pretty silly that I have to switch to a different application if I want to write normal text with typical, first-line indentation standards. Otherwise, I have to trip over an auto-indent feature that can't be turned off.

**Comment by ajbogh** [comment]
Any information on whether this feature was implemented, or maybe a workaround to disable auto-indentation? I'm working on a React project where the indentation gets confused when HTML is inserted into the js files, even though "Javascript React" is chosen as the file type. I'd rather disable auto-indentation and indent manually, or at least use same-level indents where the editor doesn't try to tab the next line in, instead it attempts to maintain the same level of indentation unless the user presses the tab key to indent further.

**Comment by aclarkxyz** [comment]
This certainly seems to be the IDE trend these days, but it is profoundly annoying: having really basic auto-indenting to the same column as the previous statement is great, but having the level jumping back & forth as you type, or presuming you want an extra one, is a pointless productivity killer. VSCode is a fantastic tool in general because it's lightweight and relatively simple - not cluttered up with a trillion features that think they know more about programming than you do - but we really need to be able to switch off some of the extra fancy stuff.

**Comment by cfjedimaster** [comment]
I'm confused -the last update supposedly adds indent stuff (https://code.visualstudio.com/updates/v1_14#_auto-indent-on-type-move-lines-and-paste) yet I still can't disable auto indent, even with editor.autoIndent:"false".

**Comment by quanticle** [comment]
editor.autoIndent controls whether the editor with indent or unindent based upon the indentation rules defined for the language. So, for example, you can define a rule for python that indents in on ":\n", and then out-dents on  "return". 

Setting editor.autoIndent to false disables that particular form of autoindentation, and then causes VSCode to revert to the older form of auto-indenting, where it keeps the next line at the previous line's level of indentation. It does not disable autoindentation entirely.

Personally, I think the flag is mis-named. I would have called it something like "editor.followIndentRules" or something similar rather than "editor.autoIndent".

**Comment by cfjedimaster** [comment]
For me, in HTML at least, it is auto-indenting. So it isn't staying at the same level. Either way - it sounds like it is a bad name as you say.

I find it a bit frustrating that *other* stuff was added in regards to indentation before being able to completely disable it.

**Comment by Kwon11** [comment]
this feature is great, but if you can't disable it then it only has to behave unexpectedly once for it to become infuriating.

**Comment by sk8terboi87** [comment]
I've updated the setting with  "editor.autoIndent": false" but still, it keeps on doing auto indent. I just switched from Sublime Text to VS Code and it is really awesome, but this issue is very annoying.
```
{
    "sublimeTextKeymap.promptV3Features": true,
    "editor.multiCursorModifier": "ctrlCmd",
    "editor.snippetSuggestions": "top",
    "editor.formatOnPaste": true,
    "git.ignoreLegacyWarning": true,
    "workbench.colorTheme": "Monokai",
    "php.suggest.basic": true,
    "files.autoSave": "afterDelay",
    "editor.autoIndent": false,
    "files.trimTrailingWhitespace": false,
    "workbench.panel.location": "bottom"
}
```

**Comment by codenamezjames** [comment]
Is there a status on this??

**Comment by auchenberg** [comment]
Question on StackOverflow has passed 750 views. https://stackoverflow.com/questions/36664324/how-to-disable-autoindent-on-enter

**Comment by Rhywun** [comment]
I don't mind the behavior where pressing Enter keeps the next line at the previous line's level of indentation. But that should only happen the first time you press Enter. I can press Enter ten times and I get nine blank lines followed by a line indented at the level of the line ten lines up. That's madness. That sort of indentation is not useful in any language that I'm aware of. The second time you press Enter without typing anything, the cursor should be positioned at column one.

**Comment by rcook** [comment]
I think this behaviour is particularly problematic with programming languages like Python and Haskell where the autoindent often (nearly always) produces invalid or semantically different code.

**Comment by mainakk** [comment]
It is also problematic with Scheme (and Scheme-like languages). It doesn't let format code in the following way:

```
(define f
  (pp 1)
  )
```

It changes it to:

```
(define f
  (pp 1)
)
```

**Comment by bertfw** [comment]
It should be possible to disable this auto-indent feature entirely, meaning that when you hit Enter the cursor is positioned at column 1 on the next line. I'm hardwired to hit tab when I know I'll need an indent, so having the editor forcibly add an indent is infuriating. There are times when keeping the previous level of indentation could be a positive thing, but these are so far and few between that using select-block-tab is just as good, making auto-indent rather pointless.

In fact I'd even say that in general, the control of anything "autoformat" should be left to the end user.

**Comment by aclarkxyz** [comment]
There should really be 3 settings for autoindent:

- nothing (rewind to column 1)
- smart (try to guess what the user wants)
- keep (indent to same position as previous line)

I can't imagine anyone would feel left out if all of these were available as configuration choices. The default can be whatever is most popular.

**Comment by thenexus00** [comment]
I am having this issue mainly with the CSS. Every time I add a new CSS line it starts indented instead of a new line. Here is just one example:
`#blog-detail .social-likes .twitter-tweet-button
								{  }
								HERE`
It actually gets pretty frustrating :(

**Comment by ZhaoShuangyin** [comment]
I also think that auto-indentation is sometimes very annoying and I hope to be able to customize the "enable" or "disable".

**Comment by thenexus00** [comment]
It is not to do with the code in terms of errors but rather the CSS format.
VSCode right now only considers the designer style format and not other 
format conditions like single line Pico structure.
I am even seeing similar issues with Flutter (Dart) as well during 
mobile apps and even if your doing standard formats.

**Comment by bertfw** [comment]
Hi Andrewriveraj, It doesn't have to do with save. Whenever hitting "Return" to create a new line it autoindents the new line even if all indenting options are disabled. This does not seem to be an issue on the Windows version, at least not in my experience - only OS X. I have not tried the Linux version.

**Comment by ajbogh** [comment]
I've come to the conclusion that prettier may be the solution to this problem. Instead of using VSCode's formatter, we should instead default to prettier.

https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode

With eslint you might use the eslint-prettier package following similar instructions as these:

https://medium.com/@sgroff04/configure-eslint-prettier-and-flow-in-vs-code-for-react-development-c9d95db07213

On particular place where I had trouble was with React development using Flowtype with the "Javascript React" layout preset. Typescript worked fine with the "Typescript React" preset but "Javascript React" broke after the first JSX code.

**Comment by roman01la** [comment]
Prettier is tied to supported languages and not necessary a tool that someone might want to use.

**Comment by peredin** [comment]
@ajbogh I fail to see how Prettier is even slightly related to the issue of auto-indentation not being disableable. Even if it wasn't "tied to supported languages" as mentioned earlier.

Prettier is used by some of the projects I work with and it is great for auto-formatting code, but it doesn't help me to _not_ have new lines indented by default.

**Comment by cfjedimaster** [comment]
@peredin Thank you - it really feels like things are straying from the original request. :)

**Comment by ajbogh** [comment]
@peredin, it is a tool that helped me in some ways with the auto-indentation. You're absolutely right that it doesn't fix the underlying problem or solve cases where prettier cannot be used, but for those that it can be used it should be noted as an alternative. Hopefully the information that I provided can help someone.

**Comment by thenexus00** [comment]
Hi, I had a mobile app deadline and got super busy.
In terms of CSS here is a quick video I just made:
http://recordit.co/ddjWLwyF7D
I got several issues with the indenting as I mentioned but this shows how Even if you point to the start of a line and paste the auto indent kicks in when it really shouldn't. I am not sure the thinking of why. If I hit enter it starts in a place I think is OK but if I decide to point to the start and paste something, thats where I want it to go. For my CSS I prefer a Pico format of one line but there are about 4/5 most commonly used CSS formats either pure or then if using something like LESS.

Visual studio seem to try and help considering only one single CSS layout method.

Further to my issues - Building Flutter / Dart apps I am encountering the same problems. The standard practises and formats developers are looking to have here which I am trying to follow Visual studio starts to be very annoying with the auto indent predictions.

Basically there are few formats of code patterns for different languages and file types. Visual studio is trying to be smart and help you but is only considering a handful of methods for each.

**Comment by DavidEGrayson** [comment]
The autoindentation in VSCode is really annoying when editing HTML, because it does not understand basic parts of HTML5 syntax, such as the fact that opening an `<li>`  tag directly inside another `<li>` tag implicitly closes the earlier tag.  Copying and pasting list elements inside such a list becomes a big chore because the auto indentation screws everything up, and I have to press Shift+Tab a few times to unindent my selection after pasting.

It's kind of shocking there is no way to turn off this annoying feature, except for clicking on "HTML" in the lower right corner, typing "plaintext", and pressing enter.

**Comment by aggieben** [comment]
I sometimes find that VSCode gets in a state where it thinks it needs to indent the wrong number of spaces (say 8, instead of 4) based on some formatting somewhere in the file.  I generally like autoindent, but it's maddening when it gets in this state.

**Comment by wbrian-onlineed** [comment]
I was enjoying being a new user of VSCode until today when i ran into this issue. Now it's just infuriating. I'm working on a codebase that has a quarter million lines using Ratliff indentation and i have to reverse VSCode's attempts to be helpful on each and every "}".

+1 on idea to have a "keep" indentation mode that just stays on the same line unless i say otherwise. My other text editor has this and it seems silly the VSCode does not.

**Comment by Neonit** [comment]
I was just about to start liking VS Code really much, but this basic feature is missing since 2016? What the hell, guys, are you serious about this?

**Comment by nate-owens** [comment]
I'd also love the ability to turn this off.  I really love VS Code, but when I'm working in PHP and mixing HTML with PHP tags, the editor gets totally confused with its indentions.

**Comment by 9527KGG** [comment]
I just want a notepad/basic experience for no reason. Thanks.

**Comment by frenchtoastbeer** [comment]
Please!?  Pretty please!?  VSCode is regarded by me as best in class.  Things work, and relatively straightforward, and that's great.  One brief and relatively small install and I get a syntax highlighting editor with a console window that knows how to color all the syntax for all the languages I use.

What isn't great, is that I can't find all the damn buttons to turn off Microsoft's addiction to moving my cursor, closing my brackets, closing my strings, finishing my sentences, or offering suggestions to name my children.  

I want to click one button labeled, "Don't pretend you know me." and forever afterward vscode simply accepts that we'll never be soulmates and stops trying to make our relationship more than it is.

Is this ever going to happen?  

Please?

**Comment by David263** [comment]
**This long Issue, lasting about 3 years, is pitiful!** I was evaluating VSCode to replace my great NoteTab Pro editor (which cannot handle Unicode), but this enforced automatic indent behavior is absolutely not acceptable. VSCode is otherwise so nearly complete and nearly what I want. But if its developers cannot respond to what users clearly feel is a bug, then the product is unusable. This Issue would move the heart of a stone, yet produces no fix. No way will I get dependent on using a product with such poor support.

**Comment by redtide** [comment]
Is Microsoft all the same that tries to make things in place of people thinking they are like idiot and do it also badly? This indentation thing is really bad and annoying and the worse thing that the best thing to do is so simple: doing nothing! Why I can't disable this shit?

**Comment by David263** [comment]
redtide, as I indicated, this is the reason I rejected VSC as my new editor. So far I have evaluated 13 editors and none is as good as my current editor for my purposes, NoteTab Pro. Unfortunately, my editor cannot handle Unicode and is no longer maintained, so I have to move to another editor. Currently, I'm trying to find workarounds that will make UltraEdit acceptable for my purposes (working on many small projects and subprojects that don't use frameworks or libraries). VSC is great in many ways, but automatic indentation that cannot be disabled is an interruption of productivity with which I cannot live. I think you are right, Microsoft cannot see a bug when it's right in front of them.

**Comment by redtide** [comment]
@David263 indeed, I tried one time to try Atom, which VSCode is a fork of AFAIK but I was lost there and I missed some feature that I'm using, which are not so much but I need those. It's hard to accept to have a good tool and not able to use it because just one stupid funky useless and bad feature like this.

**Comment by David263** [comment]
We cannot encourage a poor design by using it. Big companies have to learn not to be so arrogant. A useful design principle is "least astonishment".

**Comment by aclarkxyz** [comment]
Hey guys, maybe a little bit less hating on this thread? First of all, VSCode is doing a really good job of honouring open source principles - it's totally free, actively maintained at great expense to someone else, and with the exception of the occasional glitch, it's really good. Regarding the auto-indenting: there were a few releases in which it was incredibly annoying and broken, and I had to hard-patch some of the code each time it updated itself. But now if you switch off the 'Auto Indent' flag in the settings, most of the really intrusive behaviour stops. It's not precisely what I'd like, but it's close enough to silence my complaints.

**Comment by redtide** [comment]
> But now if you switch off the 'Auto Indent' flag in the settings, most of the really intrusive behaviour stops.

@aclarkxyz What version is that? Here it doesn't change at all.

**Comment by David263** [comment]
Alex, I apologize for the hating. It's just that it is so frustrating when any software (open source or not) is not fixed after a period of many years of complaints. I appreciate the fixes you have done, but **turning off the auto indent option was the first thing I did** the day I began to evaluate VSC. You may have fewer complaints, but there are many times that I pressed Enter during my testing and the caret did not go to the beginning of the next line as expected. At first, this astonished me. After a few times, it annoyed me. Why? Because it's a bug, and because it was reported years ago. Yes, it may be true that the worst of the intrusive behavior no longer happens (I never saw the really bad behavior, I guess), but it happens enough that I cannot adopt VSC as my new editor. You should care about this, and help get the VSC developers to fix the bug completely. Bugs should not survive one month, never mind several years, in my opinion. In my 40 year software career I always fixed bugs assigned to me completely and quickly. It is my experience that bug fixing is either a lost art or people don't care anymore, for few products these days are supported the way they should be, open source or not.

**Comment by redtide** [comment]
Well it's really stupid to set by default something instead let people set if needed, it's their usual behavior, the same reason I stopped using windows. I was kinda unsure to install their product on my system and for a while was OK, I have endured this indentation behavior too much long and now I need to switch back to the previous editor and loose time to set it up from scratch as this configuration. There is also my time and trust, not only their benevolence to share freely their s...oftware.

**Comment by Neonit** [comment]
From my personal experience I can now say that it is an exaggeration to say my productivity is constantly be lowered by this. I got used to it quite quickly. The problem is, however, that I don't want to get used to auto-indentation. It's not a time-saver in any way for me.

I can fully understand the rage about this topic as well, because I for my part just am *not able to understand* **why** this is such a big deal that stays open for **over three years** (and this makes me frustrated).

Where are the devs, who know the auto-indentation/-completion system inside-out (or at least better than the average user) and either could easily change this in minutes or could at least explain, why it is difficult or not possible, maybe contribute their knowledge at least so someone with way less insights (e.g. me) could get a starting point to have a look on their own.

I hope this is not being ignored, because having it as a force-enabled feature is some religious thing by the devs.

**Comment by aclarkxyz** [comment]
> > But now if you switch off the 'Auto Indent' flag in the settings, most of the really intrusive behaviour stops.
> 
> @aclarkxyz What version is that? Here it doesn't change at all.
> 
> Versione: 1.40.1
> Commit: [8795a98](https://github.com/microsoft/vscode/commit/8795a9889db74563ddd43eb0a897a2384129a619)
> Data: 2019-11-15T22:58:14.270Z

Two issues, I think. Now and at all times in the past, it will add an extra indent level for you under certain circumstances. Personally I do not like this (and that's the opinion of everyone else on the thread too), but I have made my peace with it: I have so few other complaints that I can learn to live with this relatively minor nuisance. And since almost every other modern editor is so much worse, I figured I'd just call this a win.

The other issue is that there were several releases where it was *severely* broken for anyone who doesn't use one of the wretched misaligned-brace styles: it would guess your indent level (wrongly), and then adjust it to something else as soon as you typed a character. It was totally unusable, and the only solution was to hack some of the source files after each release. This did eventually get fixed. If this sort of thing happened on a regular basis, I'd be full of rage, too... but fortunately it doesn't!

**Comment by David263** [comment]
Alex, I think you are remarkably, amazingly tolerant. I'm not. My software tools must work, they must work well, they must further my productivity, and they must not cause repeated frustration. If they do, I report bugs. If the organization that supports a tool does not believe in fixing bugs unless they are so extreme that the tool cannot be used, I abandon the tool and the organization, not because I want to but because I must.

In the case of editors and browsers, for some reason (as opposed, say, to javascript frameworks or string conversion tools), organizations appear to tolerate bug reports of many years' duration. This is well beyond my understanding: computers are supposed to make our life more fun, not give us frustration, and developers are supposed to fix bugs, not rationalize their continued existence. VSC is supposed to be a new product, yet it acts like a product bent over with rheumatism, hobbling around like an old man (I'm 73 years old and I can still find and fix bugs).

You and I live in different worlds so far as our philosophy of software tool maintenance is concerned. That would be fine if the stakes were not quite so high. But I will not compromise on software tool usability. Life is too short to live in frustration.

Although it is true that few of the software tools I have created have had any bugs, I have always fixed my own bugs quickly (I posted that already; you may have noticed). Why can't others fix their bugs as well as I did throughout my 40 year career? Is it laziness, or the fact that these developers leave too many bugs behind, or write unmaintainable code? There has to be a reason.

In the software industry, we implement in stages. We first create a design specification and review it, so problems can be caught early, where they are easier to fix. Then we unit test and do code reviews (I'm abbreviating) so coding bugs can also be caught early, where they are easier to fix.

I see no reason why using an Open Source approach to collaboration has to produce many unresolved bugs, or has to result in justifying the continued existence of bugs years after they have been reported.

There is something broken about how VSC was created and/or is now supported.

I'm not filled with hate, as you first supposed, and I hope you appreciate that I am presenting balanced opinions, not incivilities or anger. I'm simply explaining that something is wrong with VSC support and that I cannot adopt VSC in spite of its many reliable and well-thought-out and well implemented features.

There is just no excuse for not fixing bugs (defined as receiving bug reports from various users), especially the amazing excuse that the feature used to have an even worse bug that was fixed.

**Comment by madprops** [comment]
"There is just no excuse for not fixing bugs" 

They don't owe you anything.

I came here because I'm trying to disable autoindentation on a file, not globally.
Auto indentation is incredibly useful though.

**Comment by David263** [comment]
"They don't owe you anything"  Well, that's true. But if a great product has one particular glaring bug, one that has been reported again and again for years, but the supporters seem to refuse to fix it, why shouldn't I complain?

You are welcome to believe that autoindentation is useful. In 50 years of software engineering, I have never wanted autoindentation. The Enter key should just do a CR and LF, nothing more. If you want indentation, you should use the Tab key. I find autoindentation particularly annoying when editing text files, but it always interferes with my productivity.

That is what options or preferences are for: to accommodate different needs and ways of working. But VSC's "autoindention off" option doesn't work.

Your turn.

**Comment by madprops** [comment]
Would be nice sure, but there might be reasons this hasn't been worked on. Maybe other stuff is priority right now. Every month there are tons of updates in the changelog. Also, it's an open source project, meaning you can fix it yourself if you want.

**Comment by redtide** [comment]
> Would be nice sure, but there might be reasons this hasn't been worked on. Maybe other stuff is priority right now. Every month there are tons of updates in the changelog. Also, it's an open source project, meaning you can fix it yourself if you want.

So low priority that after 3 years and half and 50 comments on the issue nobody gave it a shit :D

**Comment by Neonit** [comment]
> Would be nice sure, but there might be reasons this hasn't been worked on. Maybe other stuff is priority right now. Every month there are tons of updates in the changelog. Also, it's an open source project, meaning you can fix it yourself if you want.

And I think it's about time to explain these reasons here and now.

**Comment by alexandrudima** [comment]
Hi everyone, I want to apologize for my silence on this issue. I personally unsubscribe from feature requests in my areas because I cannot handle the volume of comments. I am just overwhelmed by comments otherwise, and I am an incredibly slow reader and writer of comments.

We try to incorporate community feedback as much as possible into our planning, and we look at [the most popular feature requests](https://github.com/microsoft/vscode/issues?page=4&q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) when planning. Unfortunately, this issue is not in the top 100.

Please also consider that we are an open source project and it is always possible to submit a PR implementing the desired functionality. A PR, especially a well written one that can be merged immediately, will greatly expedite a feature request shipping in VS Code.

This issue is marked as a feature request and not a bug because the current code works as intended. That is, the code matches the desired intent and this issue does not identify a flaw with the code. In other words, the code correctly preserves indentation, or indents when having an open bracket, exactly as desired when we looked at how Enter should behave. We treat bugs differently than feature requests and we never have unassigned/unsubscribed bugs.

Finally, I agree that we should have an option controlling this behavior and I'm sorry for disappointing you on this issue.

**Comment by redtide** [comment]
@alexandrudima an option to disable a feature that doesn't work IS a bug, not a feature request.

**Comment by geekley** [comment]
Here's what I dislike about the current way VSCode does indentation:
```js
const app = connect()
  .use(fn);
```
If I press enter after the `;`, the next line will have the same indentation as the that line. Even after removing the indentation on this line, when I press enter again, the next line will have that indentation again. This kinda seems like a bug to me.

**Comment by alexandrudima** [comment]
This thread has captured several distinct problems... The current implementation for handling <kbd>Enter</kbd> in the editor is a bit complex and influenced by quite a number of things:
1. `onEnterRules` -- these are language specific rules. For example, this is the mechanism through which pressing <kbd>Enter</kbd> in a JSDoc comment will maintain the ` * `, which is specific to JS:
![TO_UPLOAD](https://user-images.githubusercontent.com/5047891/69721216-a7c5d600-1114-11ea-8fc6-78090c571221.gif)

2. `brackets` -- each language can define its brackets. This is the mechanism through which pressing <kbd>Enter</kbd> after an open bracket or inside a bracket pair will indent or, respectively, indent-newline-outdent:
![TO_UPLOAD](https://user-images.githubusercontent.com/5047891/69721454-2884d200-1115-11ea-94b7-9bff9f3f20a3.gif)

3. `indentationRules` -- each language can define its indentation rules. These rules are applied in a number of places, like when pasting, or moving lines. Since `indentationRules` are regex based and thus cannot be always correct, there is a setting `"editor.autoIndent"`, which can disable them.
![TO_UPLOAD](https://user-images.githubusercontent.com/5047891/69730548-925aa700-1128-11ea-83d6-8ccb533d69f9.gif)


4. if none of the above applies, the indentation is maintained on the new line.

---

Here I will try to sum up the concrete problems mentioned in this issue:

#### Pressing Enter in HTML after an element like `<p>` indents unexpectedly

![TO_UPLOAD](https://user-images.githubusercontent.com/5047891/69720727-90d2b400-1113-11ea-9f3c-730a0620cd80.gif)

In this case, the indentation is increased via `onEnterRules`. This is caused [by this specific HTML `onEnterRule`](https://github.com/microsoft/vscode/blob/949b9a25d90c13ed463ef2f87ae3f4bdd7c20fe0/extensions/html-language-features/client/src/htmlMain.ts#L149-L152).

#### Pressing Enter should insert only `\n` and nothing else

![TO_UPLOAD](https://user-images.githubusercontent.com/5047891/69720898-f58e0e80-1113-11ea-9ee6-f71f8df29ec5.gif)

In this case, the indentation is maintained. This is caused by the fallback of maintaining the current indentation on the new line.

---

There are other comments which mention problems with Clojure, JSX, Python or Haskell, but which are unfortunately imprecise, so I could not distill any clear steps on enter from them.

---

**TL;DR** - Currently, the <kbd>Enter</kbd> logic has 4 discrete "plugins" that tweak what happens. Out of the 4, only one can be disabled via `editor.autoIndent`.

The challenge is to come up with an editor option which would allow turning on/off the 4 "plugins" individually.

**Comment by David263** [comment]
alexandrudima, I do sympathize with your unsubscribing due to the volume of bug reports. But fixing the bug is a much better response to a popular bug. It gets the bug reports to stop, which is better than ignoring them.

"I agree that we should have an option controlling this behavior and I'm sorry for disappointing you on this issue."

I am very impressed with your apology; it shows humility, which is a wonderful quality.

However, I also agree with the postings that follow yours, starting with "an option to disable a feature that doesn't work IS a bug, not a feature request." 

There are many situations where the current autoindenting simply doesn't work right, and a simple CRLF is what we want Enter to represent. Even if you yourself love autoindenting, and even if it were true that it works right (which it doesn't, as documented in this thread), you really have to respect these many bug reports that indicate that lots of people want to turn off autoindentation.

How many years do users have to struggle to get a user option fixed in Open Source Microsoft land? This is not how I have ever worked on software in my 40 years as a software engineer. I appreciate your wonderful apology, but a fix would be so much more useful and satisfying. I would be able to use VSC as my editor,

**Comment by aclarkxyz** [comment]
> This thread has captured several distinct problems... The current implementation for handling Enter in the editor is a bit complex and influenced by quite a number of things:
> ...

Could it perhaps be easier to just add another top level configuration flag with a name like "Preserve same indent for new line", which would do what cutting edge IDEs from the '80s and '90s did: just keep the same indentation as the line immediately before it.

The substance of this thread is that this trivial-to-implement solution is actually a nice middle ground that a lot of people like, and has been removed from too many modern IDEs. Having no autoindenting at all (i.e. going back to column 1) is not appealing to most people... but making a prediction about what _should_ happen is problematic for a number of reasons, and a lot of people (including me) don't want that.

BTW, generally speaking, great work: I enjoy using this IDE _much_ more than anything else that's going on right now.

**Comment by alexandrudima** [comment]
@David263 I appreciate your feedback, and yes, I am honestly apologizing.

---

Here is what happened in this specific case with this issue:
1. It came in 
2. It got classified as a feature request
3. I unsubscribed
4. Nobody from the VS Code team has been subscribed to the issue and the issue has not entered top 100, so nobody from the VS Code team has read the comments that you have been writing for the past 3 years until today. 
5. I read these comments today thanks to an email from the community that came our way raising the issue.
6. I try to put out a fire that I didn't know was burning, so I am sorry about that.

---

Here are tools available to you (as part of the community) to bring things to our attention instead of getting frustrated that we are not answering. But we are not answering because maybe we are not subscribed to issues, so we don't get notified when you write a comment in a feature request that is not assigned:
1. (preferable) mention people. Pick anyone from the VS Code team and mention them like this in your comment -- e.g. `@alexandrudima`. This will immediately subscribe the person you are mentioning to the thread and make a new notification for them. 
2. create a new issue. Yes, it might appear counter intuitive, but a new issue will be looked at quite quickly, while a triaged feature request, even one with 40 upvotes might not be looked at in potentially a long time.
3. send emails (like it happened in this case).
4. ping via twitter `@code`

---

If you wouldn't mind, at this point, I would like to apologize one final time and to kindly ask that we end the meta-discussion and focus again on the reported issue around <kbd>Enter</kbd>.

**Comment by David263** [comment]
aclarkxyz, I like your suggestion because of its predictability, but I'm unclear as to whether you are opposed to having a setting that would make the Enter key simply generate a newline, with no indenting at all. In my opinion, indenting "automatically" is a very hard AI problem that will likely never be solved. There will always be cases where we want the line after an indented line not to be indented, whether in plain text or in source code.

**Comment by aclarkxyz** [comment]
> aclarkxyz, I like your suggestion because of its predictability, but I'm unclear as to whether you are opposed to having a setting that would make the Enter key simply generate a newline, with no indenting at all. In my opinion, indenting "automatically" is a very hard AI problem that will likely never be solved. There will always be cases where we want the line after an indented line not to be indented, whether in plain text or in source code.

No reason to be opposed to it at all as long as it's opt-in. I suspect that it's the least popular of the 3 options, but there isn't really a case for denying it. Obviously a lot of people are fine with _predictive autoindenting_, otherwise it would not have survived becoming the only option for so many IDEs. The thing I want is _simple autoindenting_, which is probably the 2nd most popular option. And then _no autoindenting_ for people who just want nothing to do with it...

**Comment by geekley** [comment]
> The current implementation for handling Enter in the editor is a bit complex
> The challenge is to come up with an editor option which would allow turning on/off the 4 "plugins" individually.

So it seems to me that this would be solved with a configuration (which could also be overriden as language-specific) that would apply with precedence over all other logic, like:
`"editor.newLineIndentation": "none" | "repeat" | "smart"`
Where "smart" would be the default and current behavior; "repeat" always repeats the indentation from previous line; "none" goes back to column 1.

What I mean is: maybe having a 5th "override" logic is easier to do than changing all other 4.

I'm assuming this setting shouldn't apply to pasting text as well, just to enter key.

**Comment by alexandrudima** [comment]
I really like your suggestions, here is what I am planning to do:

`editor.autoIndent` will change to be `"none" | "keep" | "brackets" | "advanced" | "full"`, so you can think of it like a "slider" becoming more and more opinionated. The explanations are:

- `"none"`: Pressing <kbd>Enter</kbd> will insert `\n` and nothing more.
- `"keep"`: Pressing <kbd>Enter</kbd> will keep the indentation of the current line.
- `"brackets"`: "keep" + the special logic around brackets `{|`, `{|}` etc.
- `"advanced"`: "brackets" + respecting `onEnterRules`
- `"full"`: "advanced" + respecting `indentationRules`

This setting will be respected in multiple places:
- when pressing <kbd>Enter</kbd>
- when pressing <kbd>Tab</kbd>
- when indenting/unindenting lines

In order to not break anyone, the auto-migration will be:
* `"editor.autoIndent": false` => `"editor.autoIndent": "advanced"`
* (default) `"editor.autoIndent": true` => `"editor.autoIndent": "full"`

**Comment by David263** [comment]
@alexandrudima  Wow! Wonderful. VSC will have the best autoindenting options of any editor. I can't wait to return to VSC.

**Comment by nate-owens** [comment]
@alexandrudima I absolutely love this solution!  Thanks for prioritizing it.

And thanks to whomever sent that email! ;)
